pkg/compact/downsample/writer.go

```diff
@@ -0,0 +1,342 @@
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/chunks"
+	"github.com/prometheus/tsdb/fileutil"
+	"github.com/prometheus/tsdb/index"
+	"github.com/prometheus/tsdb/labels"
+)
+
+type symbols map[string]struct{}
+type labelValues map[string]struct{}
+
+func (lv labelValues) add(value string) {
+	lv[value] = struct{}{}
+}
+func (lv labelValues) get(set *[]string) {
+	for value := range lv {
+		*set = append(*set, value)
+	}
+}
+
+type labelsValues map[string]labelValues
+
+func (lv labelsValues) add(labelSet labels.Labels) {
+	for _, label := range labelSet {
+		values, ok := lv[label.Name]
+		if !ok {
+			// Add new label.
+			values = labelValues{}
+			lv[label.Name] = values
+		}
+		values.add(label.Value)
+	}
+}
+
+// InstantWriter writes downsampled block to a new data block. Chunks will be written immediately in order to avoid
+// memory consumption.
+type instantWriter struct {
+	dir        string
+	tmpDir     string
+	logger     log.Logger
+	uid        ulid.ULID
+	resolution int64
+
+	symbols  symbols
+	postings []uint64
+	series   []*series
+
+	chunkWriter  tsdb.ChunkWriter
+	meta         block.Meta
+	totalChunks  uint64
+	totalSamples uint64
+}
+
+func NewWriter(dir string, l log.Logger, originMeta block.Meta, resolution int64) (*instantWriter, error) {
+	var err error
+	var chunkWriter tsdb.ChunkWriter
+
+	// Generate new block id
+	entropy := rand.New(rand.NewSource(time.Now().UnixNano()))
+	uid := ulid.MustNew(ulid.Now(), entropy)
+
+	// Populate chunk, meta and index files into temporary directory with
+	// data of all blocks.
+	dir = filepath.Join(dir, uid.String())
+	tmpDir, err := createTmpDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	chunkDir := func(dir string) string {
+		return filepath.Join(dir, block.ChunksDirname)
+	}
+
+	chunkWriter, err = chunks.NewWriter(chunkDir(tmpDir))
+	if err != nil {
+		return nil, errors.Wrap(err, "create tmp chunk instantWriter")
+	}
+
+	return &instantWriter{
+		logger:      l,
+		dir:         dir,
+		tmpDir:      tmpDir,
+		symbols:     symbols{},
+		chunkWriter: chunkWriter,
+		uid:         uid,
+		meta:        originMeta,
+		resolution:  resolution,
+	}, nil
+}
+
+func (w *instantWriter) AddSeries(s *series) error {
+	if len(s.chunks) == 0 {
+		level.Info(w.logger).Log("empty chunks happened", s.lset)
+	}
+
+	if err := w.chunkWriter.WriteChunks(s.chunks...); err != nil {
+		return errors.Wrap(err, "add series")
+	}
+
+	w.postings = append(w.postings, uint64(len(w.series)))
+	w.series = append(w.series, s)
+
+	for _, l := range s.lset {
+		w.symbols[l.Name] = struct{}{}
+		w.symbols[l.Value] = struct{}{}
+	}
+
+	w.totalChunks += uint64(len(s.chunks))
+	for i := range s.chunks {
+		chk := &s.chunks[i]
+		w.totalSamples += uint64(chk.Chunk.NumSamples())
+		chk.Chunk = nil
+	}
+
+	return nil
+}
+
+func (w *instantWriter) Flush() (ulid.ULID, error) {
+	var err error
+
+	// All the chunks have been written by this moment, can close writer.
```

Comment:
How would you know that? (: It's a caller responsibiliy to do that. I would kill this comment and comment `Flush` method above properly